test/dt_arith.c

```diff
@@ -2978,74 +3459,62 @@ test_conv_flt_1(const char *name, int run_test, hid_t src, hid_t dst)
     if (H5Tconvert(src, dst, nelmts, buf, NULL, H5P_DEFAULT) < 0)
         goto error;
 
+    /* Set pointer to matching type for hardware conversion */
+    if (FLT_FLOAT == dst_type)
+        hw_p = &hw_f;
+    else if (FLT_DOUBLE == dst_type)
+        hw_p = &hw_d;
+#if H5_SIZEOF_LONG_DOUBLE != H5_SIZEOF_DOUBLE
+    else if (FLT_LDOUBLE == dst_type)
+        hw_p = &hw_ld;
+#endif
+#ifdef H5_HAVE__FLOAT16
+    else if (FLT_FLOAT16 == dst_type)
+        hw_p = &hw_half;
+#endif
+    else
+        goto error;
+
+    /* Set convenience pointer for indexing into bytes of matching type */
+    hw = (unsigned char *)hw_p;
+
     /* Check the software results against the hardware */
     for (j = 0; j < nelmts; j++) {
-        underflow = 0;
-        hw_f      = 911.0F;
-        hw_d      = 911.0;
+        int conv_ret = -1;
+
+        hw_f = 911.0F;
+        hw_d = 911.0;
 #if H5_SIZEOF_LONG_DOUBLE != H5_SIZEOF_DOUBLE
         hw_ld = 911.0L;
 #endif
+#ifdef H5_HAVE__FLOAT16
+        hw_half = 911.0;
+#endif
 
         /* The hardware conversion */
-        /* Check for underflow when src is a "larger" float than dst.*/
-        if (FLT_FLOAT == src_type) {
```

Comment:
At this point, it seemed reasonable to rewrite this part of dt_arith by creating separate functions for each type we're converting from. Otherwise, this started to become a visual mess and future support for addition types like complex numbers will only make it worse.